src/app/application.cpp

```diff
@@ -339,7 +339,7 @@ void Application::runExternalProgram(const BitTorrent::Torrent *torrent) const
     program.replace("%C", QString::number(torrent->filesCount()));
     program.replace("%Z", QString::number(torrent->totalSize()));
     program.replace("%T", torrent->currentTracker());
-    program.replace("%I", torrent->hash().toString());
+    program.replace("%I", torrent->id().toString());
```

Comment:
This should be info hash.

Reply:
What kind of hash it should be?
Now InfoHash does not have a string representation in general form. In addition, it is practically useless for the user. All the user needs in most cases is what we now call a TorrentID.

Reply:
@glassez 

For a V2 torrent, "getting its hash" should mean "getting its full SHA256 hash". In addition, for V2 torrents, there should be a way to get its truncated hash, and for hybrid torrents, its SHA-1 hash as well.

All together:
- V1 torrents: possibility to get its SHA-1 hash; the hash is the same as its ID.
- V2 torrents: possibility to get its full SHA256 hash and its truncated hash; the latter is the same as its ID.
- hybrid torrents: possibility to get its full SHA-256 hash, its truncated SHA256 hash, and its SHA-1 hash; the ID is the same as the truncated SHA256 hash, _not_ the SHA-1 hash (this is what makes most sense given the way that `get_best()` works in libtorrent IMO)

So for any given torrent, all "IDs" are hashes, but not all hashes are IDs. Accordingly, in cases such as this there should be a distinction, and we should not replace methods that previously returned a hash with ones that will now return an ID.

Reply:
You theorize a lot without looking at the practical application...
As I mentioned, all real hashes are of little use to the end user (hell, even trackers don't give a shit about them). Why do you think a custom script needs this value? Hardly for anything other than identifying the torrent. By putting something else here, we will most likely break existing scripts that depend on it. Maybe we need to rename it (or add a clarification) in GUI.
IMO, BitTorrent v2 (namely, how its backward compatibility is embodied) it makes a mess of the terminology, so we need to deal with it very carefully (so to speak, separate the grain from the chaff).

Reply:
>What kind of hash it should be?

For bittorrent v1, it is sha1. And for bittorrent v2, I imagine we would need 2 parameters, one for sha1 and another sha256 (**not** truncated).

>All the user needs in most cases is what we now call a TorrentID.

This is not "run external program" showing to the users...

Anyway, `id()` doesn't reflect our intentions here, I would suggest having some getter function to get the string representation of sha1.

Reply:
>This is not "run external program" showing to the users...

Exactly. An external program will most likely use this to identify the torrent, and you will most likely kill it by providing a real hash here. IMO, all we need is change description of "%I" param. If you want to provide real hashes you can add another parameters.

Reply:
>IMO, all we need is change description of "%I" param. If you want to provide real hashes you can add another parameters.

The "%I" param will need updates for sure, I'm currently thinking it should stay sha1 and in other PR we add another parameter for sha256 .

But this is not the main point of this thread. I'm saying `torrent->id()` feels wrong here, it is expected to see `sha1()` or some other specific getter for SHA1 hash (and not the truncated torrent id).

Reply:
Wait... I think there are merits of switching to torrent id, such as avoid breaking for existing users.
Ok lets go this way, but please add a TODO somewhere stating the GUI & webui label for this parameter needs to update.

Reply:
>I think there are merits of switching to torrent id, such as avoid breaking for existing users.

Yes, it is.
>Ok lets go this way, but please add a TODO somewhere stating the GUI & webui label for this parameter needs to update.

Ok.

Reply:
>You theorize a lot without looking at the practical application...
>As I mentioned, all real hashes are of little use to the end user (hell, even trackers don't give a shit about them). Why do you think a custom script needs this value? Hardly for anything other than identifying the torrent. By putting something else here, we will most likely break existing scripts that depend on it. Maybe we need to rename it (or add a clarification) in GUI.

What? It's quite obvious to me that the ability to extract all types of hashes is useful. Maybe you don't have a use case for it, but I'm sure plenty other users have. Until now we were able to get away with a simple implementation. Now it is clear we need 3 different parameters, like I said above. For backwards compat, `%I` should be used for the same purpose as before (SHA1 hash of V1 torrents).

A good theoretical foundation is essential for a correct implementation. In this comment, Chocobo forgot an important detail (or maybe he intentionally left it out, but like you said, it is important to be very precise here and not leave anything implicit/open to interpretation):

> For bittorrent v1, it is sha1. And for bittorrent v2, I imagine we would need 2 parameters, one for sha1 and another sha256 (**not** truncated).

Not all V2 torrents have a SHA1 hash - only the hybrid ones do.

Here is a more concrete proposal:

- `%I`: return SHA1 hash (in case of V1 torrents and V2 hybrids; empty string in case of pure V2)
- `%X`: return SHA2 hash (in case of pure or hybrid V2; empty string in case of V1)
- `%Y`: return truncated SHA2 hash (in case of pure or hybrid V2; empty string in case of V1)

`X` and `Y` are placeholders, feel free to choose other letters (I also didn't bother to check if they are actually already in use for something else).

Also, when each case does not apply, perhaps it is better to return some value like `N/A` instead of the empty string.

Reply:
@FranciscoPombal 
I'm sorry, I won't comment on your words in detail, as always there are a lot of them. Some of them are obvious (about WHAT else we need to do in connection with the addition of BitTorrent v2 support), the other part requires thoughtful reflection and discussion (about HOW we should do THAT). But I insist on the following:
1. THIS batch of changes is quite complete and it has achieved its goals;
2. it DOES NOT affect the application interface, neither the user interface, nor any other (if you haven't already understood, then all hashes are displayed exactly the same as in the current master, I just changed the internal name of some entities), so there are no regressions here that you could insist on fixing.
3. Everything else deserves a separate topic for discussion (separate PR for implementation).

Reply:
> Here is a more concrete proposal:
> 
>     * `%I`: return SHA1 hash (in case of V1 torrents and V2 hybrids; empty string in case of pure V2)
> 
>     * `%X`: return SHA2 hash (in case of pure or hybrid V2; empty string in case of V1)
> 
>     * `%Y`: return truncated SHA2 hash (in case of pure or hybrid V2; empty string in case of V1)
> 
> 
> `X` and `Y` are placeholders, feel free to choose other letters (I also didn't bother to check if they are actually already in use for something else).
> 
> Also, when each case does not apply, perhaps it is better to return some value like `N/A` instead of the empty string.

This looks extremely inconvenient for automated processing.

Reply:
@glassez 

This:

> Ok lets go this way, but please add a TODO somewhere stating the GUI & webui label for this parameter needs to update.

contradicts this:

> it DOES NOT affect the application interface, neither the user interface, nor any other 

- Before this change: User expects `%I` to return infohash, and it does so (for V1 torrents which are the only supported ones thus far)
- After this change: User gets infohash from  `%I` for V1 torrents, but **not** from V2 torrents; in the case of V2 torrents, they instead get the torrent identifier (truncated SHA256).

This is clearly inconsistent.

In the BitTorrent V2 spec, there is clearly a distinction between the V2 info-hash and the concept of a "torrent identifier" (https://www.bittorrent.org/beps/bep_0052.html, emphasis mine):

> For **some uses as torrent identifier** it [the info-hash] is **truncated** to 20 bytes.

To not break previous expectations and functionality, the `%I` should either return the full hashes for both V1 and V2, or return V1 except in case of pure V2 in which it would return the empty string (as suggested in https://github.com/qbittorrent/qBittorrent/pull/14503#discussion_r591719935 - in retrospect, while this latter alternative maintains full backwards compat, it is worse that the alternative presented in this comment, of returning full hashes in both cases).

> This looks extremely inconvenient for automated processing.

I'm all ears for alternatives (in a separate discussion, like you suggest). But if a user wants a fairly general command-line for their external program that works for both V1 and V2 use cases, it seems inevitable that at some point they will have to deal with "null"/"empty" values...

Reply:
>After this change: User gets infohash from %I for V1 torrents, but not from V2 torrents; in the case of V2 torrents, they instead get the torrent identifier (truncated SHA256)..

Just try to compile current master with libtorrent 2.